### PR TITLE
Feature/improve class scanning epmdj 10603

### PR DESCRIPTION
**Pull request by epamrsa**

### CODE REVIEW

File: test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    fun scanClasses(uris: Set<URI>) = uris.fold(0, ::addClasses).apply { classPathScanner.transferBuffer() }
+
+    fun scanClasses() = scanClasses(scanClassLoadersURIs())
+
+    fun scanClasses(additionalPaths: List<String>) = additionalPaths.run {
+        scanClasses(scanClassLoadersURIs() + this.map(::File).filter(File::exists).map(File::toURI))
+    }
+
+    private fun addClassLoaderWithParents(loaders: MutableSet<ClassLoader>, classloader: ClassLoader) = loaders.apply {
```
Comment (RomanDavlyatshin) [inline test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt:55]: tbd - mutating param

File: test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            logger?.error(it) { "ClassLoadersScanner: error retrieving classpath URIs from classloader $cl" }
+        }
+    }
+
+    private fun addClasses(count: Int, uri: URI) = count + classPathScanner.scanURI(uri)
+
+    private fun getSystemClassPath() = System.getProperty("java.class.path").split(File.pathSeparator).map(::File)
+        .filter(File::exists).map(File::toURI)
+
+    private fun normalizeURIs(uris: Set<URI>) = mutableSetOf<URI>().apply {
```
Comment (RomanDavlyatshin) [inline test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt:84]: I suggest implementing unit test, as function seems to handle lots of scenarios

File: test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassPathScanner.kt [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+private const val JAR_BUFFER_SIZE = 256 * 1024
+
+class ClassPathScanner(
+    private val packagePrefixes: List<String>,
+    private val classesBufferSize: Int,
+    private val logger: Logger?,
+    private val transfer: (Set<ClassSource>) -> Unit
+) {
+
+    private val scannedJarFiles = mutableSetOf<String>()
```
Comment (RomanDavlyatshin) [inline test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassPathScanner.kt:41]: tbd - state and mutability

File: test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    fun scanClassLoadersURIs(classloaders: Set<ClassLoader>) = classloaders
+        .fold(getSystemClassPath().toMutableSet(), ::addClassLoaderURIs).let(::normalizeURIs)
+
+    fun scanClassLoadersURIs() = scanClassLoadersURIs(scanClassLoaders())
+
+    fun scanClasses(uris: Set<URI>) = uris.fold(0, ::addClasses).apply { classPathScanner.transferBuffer() }
+
+    fun scanClasses() = scanClasses(scanClassLoadersURIs())
+
+    fun scanClasses(additionalPaths: List<String>) = additionalPaths.run {
```
[the remark this answers is not in the document]
Comment (RomanDavlyatshin) [reply test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt:51]: Can we pass `additionalPaths` to constructor, defaulting to emptyList?

File: test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import java.net.URL
+import java.net.URLClassLoader
+import com.epam.drill.logger.api.Logger
+import com.epam.drill.common.classloading.ClassSource
+
+class ClassLoadersScanner(
+    packagePrefixes: List<String>,
+    classesBufferSize: Int = 50,
+    private val logger: Logger? = null,
+    transfer: (Set<ClassSource>) -> Unit
```
Comment (RomanDavlyatshin) [inline test2code-agent/src/main/kotlin/com/epam/drill/plugins/test2code/classloading/ClassLoadersScanner.kt:29]: tbd - transfer responsiblity

**Review by RomanDavlyatshin — CHANGES_REQUESTED** [review]
Need to discuss possible changes

**Comment by epamrsa** [comment]
Discussed on code review session.

**Pull request merged**